### PR TITLE
WIP: booter for fit images.

### PR DESCRIPTION
**Pull request by rminnich**
This is just a first pass. Comments welcome. In particular, I think it makes
more sense to copy the kernel/initramfs to files in /tmp and use the file load
instead of trying to use kexec classic? comments?

Signed-off-by: Ronald G Minnich <rminnich@gmail.com>

### CODE REVIEW

File: pkg/boot/fit/fit.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	// Find the kernel name and its needed data area.
+	if verbose {
+		log.Printf("Find kernel %s", i.KernelName)
+	}
+	kb, err := i.Root.Bytes(i.KernelName, "data")
+	if err != nil {
+		return err
+	}
+	i.mem.Segments.Insert(kexec.NewSegment(kb, kexec.Range{Start: 0x100000, Size: uint(len(kb))}))
```
Comment (hugelgupf) [inline pkg/boot/fit/fit.go:73]: I thought a FIT would tell you the start address?

File: pkg/boot/fit/fit.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	if len(i.InitRAMFS) > 0 {
+		if verbose {
+			log.Printf("Find initramfs %s", i.InitRAMFS)
+		}
+		ib, err := i.Root.Bytes(i.InitRAMFS, "data")
+		if err != nil {
+			return err
+		}
+		i.mem.Segments.Insert(kexec.NewSegment(kb, kexec.Range{Start: 32 * 0x100000, Size: uint(len(ib))}))
```
Comment (hugelgupf) [inline pkg/boot/fit/fit.go:89]: ditto

File: pkg/boot/fit/fit.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	i.entryPoint = 0x100000
```
Comment (hugelgupf) [inline pkg/boot/fit/fit.go:91]: and ditto

File: pkg/boot/fit/fit.go (hunk 1 above)
Comment (hugelgupf) [reply pkg/boot/fit/fit.go:73]: I'll note that we don't do this well right now -- this does not check that the segment is not already taken, iirc.

File: pkg/boot/fit/fit.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	kb, err := i.Root.Bytes(i.KernelName, "data")
+	if err != nil {
+		return err
+	}
+	i.mem.Segments.Insert(kexec.NewSegment(kb, kexec.Range{Start: 0x100000, Size: uint(len(kb))}))
+	// Technically, we can have both an initramfs and a rootfs.
+	// The current question: how do we provided the rootfs from a file
+	// to the kernel? We do not know.
+	if len(i.RootFS) > 0 {
+		return fmt.Errorf("No way to provide a rootfs yet")
```
Comment (hugelgupf) [inline pkg/boot/fit/fit.go:78]: lower case error messages

(but as long as no rootfs can be provided, just remove the struct member?)

File: pkg/boot/fit/fit.go (hunk 2 above)
Comment (hugelgupf) [reply pkg/boot/fit/fit.go:89]: these properties? https://github.com/platinasystems/goes/blob/47b24c8c09e1741850989e581a2b9f0992d6e8d9/internal/kexec/linux_file_load.go#L41

File: cmds/boot/fitboot/main.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+func main() {
+	flag.Parse()
+	if *debug {
+		v = log.Printf
+	}
+	if len(flag.Args()) != 2 {
+		log.Fatal("Usage: fitboot uimage kernel")
+	}
+	f, err := fit.New(flag.Args()[0])
```
Comment (hugelgupf) [inline cmds/boot/fitboot/main.go:34]: can you just add fit support to the kexec command we already have?

**Comment by rminnich** [comment]
we're redoing it.

**Pull request closed without merging**